[BUG] read_excel: fixes handling of multi index header and other corner cases

PL-SalvadorFandino opened the pull request:
- [x] closes #58898
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [x] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

salva committed:
TST: read_excel: test multiindex headers with skip rows between.

read_excel function has several bugs regarding how it handles
combinations of header, skiprows and index_col arguments.

The tests here showcase some of them.

salva committed:
[BUG] excel _parse_sheet revamp

The logic in that method was not handling correctly all the possible
combinations of skiprows, header and index_col arguments.

Specifically:

- it was not able to handle correctly multi index header
  with holes (for instance, `header=[0,2]`.

- multi index header and skiprows given as lists.

- forward filling index columns and skiprows gigen as lists.

- inconsistences processing one-element list arguments (for instance,
  `header=1` and `header=[1]` or `index_col=0` and `index_col=[0]` where
  handled differently).

The logic has been revamped, because it was not possible to fix all
the errors with local changes.

The mayor challenge was handling skiprows as a list, as it may remove
rows at any place (before, between or after header(s), index names and
data). Also, header row indexes reference rows **after** removing
skiprows.

To handle that we use an intermediate mapping `ixmap` which goes from
the row indixes with skiprows removed to the row indixes in `data`.

Finally, let me add that IMO, most of the functionality of
_parse_sheet should be moved down into TextParser... but that's work
for another day!

PL-SalvadorFandino commented:
Hi, can't see why check `pre-commit.ci - pr` is failing!

Aloqeely commented:
pre-commit.ci autofix

Aloqeely commented:
Looks like I don't have the permissions to do it, but you should be able to autofix by commenting `pre-commit.ci autofix`

PL-SalvadorFandino commented:
pre-commit.ci autofix

salva committed:
read_excel: call skiprows callback just once per row.

For callable skiprows arguments, the callback was being called several
times in every row.

Now, it is called just once and the result cached.

PL-SalvadorFandino commented on pandas/io/excel/_base.py:
```diff
@@ -817,95 +817,115 @@ def _parse_sheet(
+                skiprows_set = set(cast(Sequence[int], skiprows))
+                ixmap = [ix for ix, _ in enumerate(data) if ix not in skiprows_set]
+            elif callable(skiprows):
+                ixmap = [ix for ix, _ in enumerate(data) if not skiprows(ix)]
```
Ok, now when skiprows is a callable, it is called on every row and the result cached so that effectively, it is only called once per row.

PL-SalvadorFandino commented:
I have rebased my branch on top of the current  "main" branch, because I was having some issues with the test suite unrelated to my code.

PL-SalvadorFandino commented:
Now, ixmap is not coerced into a list in order to reduce the overhead when `skiprows` is None or an integer.

PL-SalvadorFandino commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

PL-SalvadorFandino commented:
pre-commit.ci autofix

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

PL-SalvadorFandino commented:
> This pull request is stale because it has been open for thirty days with no activity. Please [update](https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#updating-your-pull-request) and respond to this comment if you're still interested in working on this.

not stale!

rhshadrach commented on pandas/io/excel/_base.py:
```diff
@@ -817,95 +817,122 @@ def _parse_sheet(
+
+                if len(header_list) > 1:
+                    if index_col_has_names:
+                        raise ValueError(
+                            "named index_col can not be used together "
+                            "with multi-index header"
+                        )
```
Is this restriction new?

PL-SalvadorFandino replied:
No, it was already there (line 888 of the original source file).

rhshadrach commented on pandas/io/excel/_base.py:
```diff
@@ -817,95 +817,122 @@ def _parse_sheet(
+
+                        if has_index:
+                            header_name, _ = pop_header_name(
+                                data[row1], sorted(index_col_set)
```
Why sort?

rhshadrach commented on pandas/io/excel/_base.py:
```diff
@@ -817,95 +817,122 @@ def _parse_sheet(
+                            header_name, _ = pop_header_name(
+                                data[row1], sorted(index_col_set)
+                            )
+                            if header_name:
```
Is the case header_name being None hit by tests?

rhshadrach commented on pandas/tests/io/excel/test_readers.py:
```diff
@@ -1759,3 +1759,53 @@ def test_corrupt_files_closed(self, engine, tmp_excel):
                 pass
+
+    def test_mi_header_skiprows1(self, engine, read_ext):
+        if engine is None and read_ext == ".xlsx":
```
Can you add references to the issue as a comment in the first line of each test:

    # GH#58898

rhshadrach commented on pandas/tests/io/excel/test_readers.py:
```diff
@@ -1759,3 +1759,53 @@ def test_corrupt_files_closed(self, engine, tmp_excel):
+
+    def test_mi_header_skiprows1(self, engine, read_ext):
+        if engine is None and read_ext == ".xlsx":
+            with open("test_mi_holes.xlsx", "rb") as f:
+                expected = pd.read_excel(
+                    f, sheet_name="expected", header=[0, 1], index_col=[0, 1]
+                )
```
Instead of adding files to the data/ directory, is it possible to generate these sheets on-the-fly? E.g.

```python
df = pd.DataFrame(...)
df.to_excel(tmp_excel, sheet_name="test", index=False)
```

`tmp_excel` here is a pytest fixture.

marcel-goldschen-ohm commented:
It would be great if a fix for this issue could be implemented quickly as it is fundamental to properly reading in data from excel files. I would suggest that this should be VERY HIGH PRIORITY.

marcel-goldschen-ohm commented:
Note, I just tried [https://github.com/PL-SalvadorFandino/pandas/tree/holes](https://github.com/PL-SalvadorFandino/pandas/tree/holes) and it does NOT solve the following issue for me:

Consider an excel sheet with the following content (three multiindex header rows with a blank row separating rows 1 and 3):
```
Group, A, A, B, B
ID, 0, 1, 0, 1
 , , , , 
X, Y, Y, Y, Y
100, 0, 1, 2, 3
200, 4, 5, 6, 7
```

When reading in the above excel sheet using `pandas.read_excel` I get mangled data. For example,
```python
pd.read_excel(..., index_col=0, header=[0,1,3], skiprows = lambda row: row not in [0,1,3,4,5])
```
gives the following output (index_col missing last data row, and other columns missing last header row):
```
Group, A, A, B, B
ID, 0, 1, 0, 1
X, 0, 1, 2, 3
200, 4, 5, 6, 7
```

Thinking that header=[...] indices might refer to row indices after removing skipped rows, I tried:
```python
pd.read_excel(..., index_col=0, header=[0,1,2], skiprows = lambda row: row not in [0,1,3,4,5])
```
which gives the following output (index_col missing last header row, and other columns ok):
```
Group, A, A, B, B
ID, 0, 1, 0, 1
 , Y, Y, Y, Y
100, 0, 1, 2, 3
200, 4, 5, 6, 7
```

Note that if the index_col is ignored, then everything works if one assumes header=[...] denotes row indices after removing skipped rows (I find this HUGELY non-intuitive):
```python
pd.read_excel(..., header=[0,1,2], skiprows = lambda row: row not in [0,1,3,4,5])
```
which gives the correct output (albeit the desired index column is now a regular column):
```
Group, A, A, B, B
ID, 0, 1, 0, 1
X, Y, Y, Y, Y
100, 0, 1, 2, 3
200, 4, 5, 6, 7
```

I have the suspicion that index_col and other columns may be treating the indices passed in header=[...] differently (e.g., row indices as in the excel sheet or row indices after removing skipped rows)? If so, I STRONGLY suggest that all row indices should be relative to the excel file, not the remaining rows after extracting the desired rows.

mroeschke commented:
Thanks for the pull request, but it appears to have gone stale. If interested in continuing, please merge in the main branch, address any review comments and/or failing tests, and we can reopen.
